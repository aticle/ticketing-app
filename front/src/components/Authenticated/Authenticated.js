// @flow
import React, { Component } from 'react';
import { connect, type Dispatch } from 'react-redux';
import { withRouter, RouterHistory } from 'react-router-dom';
import { getJwt } from '../../helpers/jwt';
import { type UserLogin, type User, getUser } from '../../actions/authAction';

type Props = {
    history: RouterHistory,
    getUser: User
};
type State = {
    user: UserLogin | null
};

class Authenticated extends Component<Props, State> {
    state = {
        user: undefined
    }

    componentDidMount() {
        this.getUser();
    }

    getUser = () => {
        const jwt = getJwt();

        if (!jwt) return this.setState({ user: null });

        this.props.getUser().then(user => this.setState({ user }));
    }

    render() {
        const { user } = this.state;

        if (user === undefined) {
            return (
                <div>Loading...</div>
            );
        }

        if (user === null) this.props.history.push('/login');

        return user;
    }
}

const mapStateToProps = () => { };
const mapDispatchToProps = (dispatch: Dispatch) => {
    return {
        getUser: async (jwt: string, history: RouterHistory) => await getUser(jwt, history)(dispatch)
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Authenticated));
