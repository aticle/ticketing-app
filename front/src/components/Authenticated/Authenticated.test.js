import React from 'react';
import { shallow } from 'enzyme';
import Authenticated, { mapStateToProps, mapDispatchToProps } from './Authenticated';

const props = {
    history: { push: jest.fn() },
    getUser: jest.fn(),
    auth: {
        isAuthenticated: undefined,
        user: undefined
    }
};
const auth = shallow(<Authenticated.WrappedComponent {...props} />);
const authI = auth.instance();

describe('<Authenticated />', () => {
    it('renders correctly when authentication not defined', () => {
        expect(auth).toMatchSnapshot();
    });

    it('redirects to login if not authenticated', () => {
        const props = {
            history: { push: jest.fn() },
            getUser: jest.fn(),
            auth: {
                isAuthenticated: false,
                user: {}
            }
        };
        const auth = shallow(<Authenticated.WrappedComponent {...props} />);
        const authI = auth.instance();

        expect(authI.props.history.push.mock.calls[0][0]).toBe('/login');
    });

    it('renders correctly when authenticated', () => {
        const props = {
            history: { push: jest.fn() },
            getUser: jest.fn(),
            auth: {
                isAuthenticated: true,
                user: {}
            }
        };
        const auth = shallow(<Authenticated.WrappedComponent {...props} />);

        expect(auth).toMatchSnapshot();
    });

    it('mapDispatchToProps', () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).getUser(authI.props.history);

        expect(authI.props.getUser.mock.calls[0][0]).toEqual(authI.props.history);
    });

    it('mapStateToProps', () => {
        mapStateToProps({ auth: props.auth });

        expect(authI.props.auth).toEqual(props.auth);
    });
});
