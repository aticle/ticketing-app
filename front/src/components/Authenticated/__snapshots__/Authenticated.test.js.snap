// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authenticated /> renders correctly when authenticated 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Authenticated
    auth={
      Object {
        "isAuthenticated": true,
        "user": Object {},
      }
    }
    getUser={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "push": [MockFunction],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [
    undefined,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<Authenticated /> renders correctly when authentication not defined 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Authenticated
    auth={
      Object {
        "isAuthenticated": undefined,
        "user": undefined,
      }
    }
    getUser={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "push": [MockFunction],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Loading...",
    },
    "ref": null,
    "rendered": "Loading...",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading...",
      },
      "ref": null,
      "rendered": "Loading...",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
